eventuals/transformer.h

```diff
@@ -0,0 +1,390 @@
+    void Fail(std::exception_ptr exception) {
+      (*fail_)(std::move(exception));
+    }
+
+    void Stop() {
+      (*stop_)();
+    }
+
+    void Ended() {
+      (*ended_)();
+    }
+
+    // Already registered in 'adapted_'
+    void Register(Interrupt&) {}
+
+    Callback<TypeErasedStream&>* begin_;
+    Callback<std::exception_ptr>* fail_;
+    Callback<>* stop_;
+    Callback<To_&&>* body_;
+    Callback<>* ended_;
+  };
+
+  HeapTransformer(E_ e)
+    : adapted_(
+        std::move(e).template k<From_>(
+            Adaptor{&begin_, &fail_, &stop_, &body_, &ended_})) {}
+
+  void Body(
+      From_&& arg,
+      Interrupt& interrupt,
+      Callback<TypeErasedStream&>&& begin,
+      Callback<std::exception_ptr>&& fail,
+      Callback<>&& stop,
+      Callback<To_&&>&& body,
+      Callback<>&& ended) {
+    begin_ = std::move(begin);
+    fail_ = std::move(fail);
+    stop_ = std::move(stop);
+    body_ = std::move(body);
+    ended_ = std::move(ended);
+
+    adapted_.Register(interrupt);
+
+    adapted_.Body(std::move(arg));
+  }
+
+  void Fail(
+      Interrupt& interrupt,
+      std::exception_ptr&& fail_exception,
+      Callback<TypeErasedStream&>&& begin,
+      Callback<std::exception_ptr>&& fail,
+      Callback<>&& stop,
+      Callback<To_&&>&& body,
+      Callback<>&& ended) {
+    begin_ = std::move(begin);
+    fail_ = std::move(fail);
+    stop_ = std::move(stop);
+    body_ = std::move(body);
+    ended_ = std::move(ended);
+
+    adapted_.Register(interrupt);
+
+    adapted_.Fail(std::move(fail_exception));
+  }
+
+  void Stop(
+      Interrupt& interrupt,
+      Callback<TypeErasedStream&>&& begin,
+      Callback<std::exception_ptr>&& fail,
+      Callback<>&& stop,
+      Callback<To_&&>&& body,
+      Callback<>&& ended) {
+    begin_ = std::move(begin);
+    fail_ = std::move(fail);
+    stop_ = std::move(stop);
+    body_ = std::move(body);
+    ended_ = std::move(ended);
+
+    adapted_.Register(interrupt);
+
+    adapted_.Stop();
+  }
+
+  Callback<TypeErasedStream&> begin_;
+  Callback<std::exception_ptr> fail_;
+  Callback<> stop_;
+  Callback<To_&&> body_;
+  Callback<> ended_;
+
+  using Adapted_ = decltype(std::declval<E_>().template k<From_>(
+      std::declval<Adaptor>()));
+
+  Adapted_ adapted_;
+};
+
+////////////////////////////////////////////////////////////////////////
+
+struct _TransformerEraser {
+  enum class Action {
+    Body = 0,
+    Fail = 1,
+    Stop = 2,
+  };
+
+  template <typename K_, typename From_, typename To_>
+  struct Continuation {
+    void Begin(TypeErasedStream& stream) {
+      k_.Begin(stream);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      std::exception_ptr exception;
+
+      if constexpr (sizeof...(args) > 0) {
+        exception = std::make_exception_ptr(
+            std::forward<decltype(args)>(args)...);
+      } else {
+        exception = std::make_exception_ptr(
+            std::runtime_error("empty error"));
+      }
```

Comment:
After thinking about this more, let's just make this a compile time error.
```suggestion
      static_assert(sizeof...(args) > 0, "Transformer expects Fail() to be called with an argument");
      auto exception = std::make_exception_ptr(
          std::forward<decltype(args)>(args)...);
```